packages/protocol/contracts/L1/TaikoData.sol

```diff
@@ -69,46 +100,36 @@ library TaikoData {
         uint256 deposit;
         address proposer;
         uint64 proposedAt;
-    }
-
-    // 3 + n slots
-    struct ForkChoice {
-        bytes32 blockHash;
-        address prover;
-        uint64 provenAt;
+        uint32 nextForkChoiceId;
     }
 
     // This struct takes 9 slots.
     struct State {
-        // some blocks' hashes won't be persisted,
-        // only the latest one if verified in a batch
-        mapping(uint256 blockId => bytes32 blockHash) l2Hashes;
-        mapping(uint256 blockId => ProposedBlock proposedBlock) proposedBlocks;
+        mapping(uint256 blockId => ProposedBlock) proposedBlocks;
         // solhint-disable-next-line max-line-length
-        mapping(uint256 blockId => mapping(bytes32 parentHash => ForkChoice forkChoice)) forkChoices;
+        mapping(uint256 blockId => mapping(bytes32 parentHash => uint256 forkChoiceId)) forkChoiceIds;
+        mapping(uint256 blockId => mapping(uint256 index => ForkChoice)) forkChoices;
         // solhint-disable-next-line max-line-length
-        mapping(address proposerAddress => mapping(uint256 commitSlot => bytes32 commitHash)) commits;
-        // solhint-disable-next-line max-line-length
-        mapping(address prover => uint256 outstandingReward) balances;
+        mapping(uint256 blockNumber => ChainData) l2ChainDatas;
```

Comment:
I think there could be some potential gas savings as well if it's done like that because it'll allow the chaindata to be accessible for a block without actually copying the data out of the block and into the separate mapping, simply by storing the correct forkchoiceid in the block that redirects to the values stored by the prover in `forkChoices`. That's the main idea, I still have to work things out a bit more to know for sure.

Reply:
Done.